In [ ]:
import pandas as pd
import numpy as np
import regex as re

In [ ]:
# Display all columns in data frames
pd.set_option('display.max_columns', None)

In [ ]:
# Read the characteristic data into a dataframe called df
df = pd.read_csv('mp_source_data.csv')
# Read the medical data into a dataframe called med_df
med_df = pd.read_csv('mp_source_data_medical.csv')

In [ ]:
# Check the top of the df dataframe
df.head(1)

Pet Name            Pet Breed  \
0  7 Kittens Litter: Henry (Courtesy Post)- no lo...  American Short Hair   

  Date Pet Entered Your Care  Length of Stay (Days) Adoption/Foster Date  \
0                  6/18/2022                     18                  NaN   

  Last Status Change     Current Status Pet Type Pet Age  \
0          21-Nov-22  Returned to Owner      Cat  Kitten   

  Pet Owner's Cell Number Pet Owner's Work Phone Pet Owner's Home Number  \
0                     NaN                    NaN                     NaN   

  Petstablished ID    Size Weight         Color Shedding Coat Length  \
0        PS1450793  Medium    NaN  Black, White      NaN         NaN   

  Temperament Breed Type Current Foster/Adopter  \
0         NaN        NaN                    NaN   

  Date Placed in Current Location Microchip ID Microchip Manufacturer  \
0                             NaN          NaN                    NaN   

  Adoption Fee Internal ID  Animal Control ID Adoption Application Visible  \
0           75         NaN                NaN                    invisible   

  Foster Application Visible Last updated at Gender  \
0                  invisible       21-Nov-22   Male   

  Previous Owner Information Acquired By Date of Birth       Age in Years  \
0                        NaN         NaN     4/20/2022  1 years, 8 months   

  Declawed Status Coat Pattern Events Attendance Is Mix Shots up to date?  \
0             NaN          NaN               NaN    Yes               YES   

  Spayed/Neutered? Hypoallergenic? Housebroken? Special Need? Needs Foster?  \
0         Not Sure        Not Sure          YES      Not Sure            NO   

  Gets along with Cats? Gets along with Dogs? Gets along with Kids?  \
0              Not Sure              Not Sure              Not Sure   

                                         Description  \
0  <p>Henry is one of seven kitty siblings (b. 4/...   

                                      Internal Notes  \
0  Reason Rehoming:Fostering them until finding h...   

  Behavioral Tracking Notes Has Finalized Application?  \
0                       NaN                         NO   

  Where was pet originally found Date Pet Entered Your Care.1  \
0                            NaN                    6/18/2022   

  Rabies Issue Date Additional Comments  Owner requested euthanasia  \
0               NaN                 NaN                         NaN   

   Additional Internal Comments Type of Intake Parents  \
0                           NaN          Other     NaN   

                                            Siblings Euthanasia Date  \
0  7 Kittens Litter: Pumpkin (Courtesy Post)  PS1...             NaN   

  Euthanasia Reason Date Transferred Transferred From Transferred To  \
0               NaN              NaN              NaN            NaN   

  Transfer Reason Deceased Date Deceased Reason Cross Post Organization Name  \
0             NaN           NaN             NaN                          NaN   

   Expected Euthanasia Date Transferred out location Transferred in  
0                       NaN                      NaN            NaN

In [ ]:
# Replace the spaces with _
df.columns = df.columns.str.replace(' ', '_')
# Change column text to lowercase
df.columns = df.columns.str.lower()
# Replace ' with empty string
df.columns = df.columns.str.replace("'", "")
# Replace special characters with empty string
df.columns = df.columns.str.replace(r'[\?\(\)]', '', regex = True)
# Replace / with _
df.columns = df.columns.str.replace('/', '_', regex = True)
# Replace .1 with _1
df.columns = df.columns.str.replace('.1', '_1', regex = True)

In [ ]:
# Check top of data frame
df.head(1)

pet_name            pet_breed  \
0  7 Kittens Litter: Henry (Courtesy Post)- no lo...  American Short Hair   

  date_pet_entered_your_care  length_of_stay_days adoption_foster_date  \
0                  6/18/2022                   18                  NaN   

  last_status_change     current_status pet_type pet_age  \
0          21-Nov-22  Returned to Owner      Cat  Kitten   

  pet_owners_cell_number pet_owners_work_phone pet_owners_home_number  \
0                    NaN                   NaN                    NaN   

  petstablished_id    size weight         color shedding coat_length  \
0        PS1450793  Medium    NaN  Black, White      NaN         NaN   

  temperament breed_type current_foster_adopter  \
0         NaN        NaN                    NaN   

  date_placed_in_current_location microchip_id microchip_manufacturer  \
0                             NaN          NaN                    NaN   

  adoption_fee internal_id  animal_control_id adoption_application_visible  \
0           75         NaN                NaN                    invisible   

  foster_application_visible last_updated_at gender  \
0                  invisible       21-Nov-22   Male   

  previous_owner_information acquired_by date_of_birth       age_in_years  \
0                        NaN         NaN     4/20/2022  1 years, 8 months   

  declawed_status coat_pattern events_attendance is_mix shots_up_to_date  \
0             NaN          NaN               NaN    Yes              YES   

  spayed_neutered hypoallergenic housebroken special_need needs_foster  \
0        Not Sure       Not Sure         YES     Not Sure           NO   

  gets_along_with_cats gets_along_with_dogs gets_along_with_kids  \
0             Not Sure             Not Sure             Not Sure   

                                         description  \
0  <p>Henry is one of seven kitty siblings (b. 4/...   

                                      internal_notes  \
0  Reason Rehoming:Fostering them until finding h...   

  behavioral_tracking_notes has_finalized_application  \
0                       NaN                        NO   

  where_was_pet_originally_found date_pet_entered_your_care_1  \
0                            NaN                    6/18/2022   

  rabies_issue_date additional_comments  owner_requested_euthanasia  \
0               NaN                 NaN                         NaN   

   additional_internal_comments type_of_intake parents  \
0                           NaN          Other     NaN   

                                            siblings euthanasia_date  \
0  7 Kittens Litter: Pumpkin (Courtesy Post)  PS1...             NaN   

  euthanasia_reason date_transferred transferred_from transferred_to  \
0               NaN              NaN              NaN            NaN   

  transfer_reason deceased_date deceased_reason cross_post_organization_name  \
0             NaN           NaN             NaN                          NaN   

   expected_euthanasia_date transferred_out_location transferred_in  
0                       NaN                      NaN            NaN

In [ ]:
# Drop personally identifiable information
df = df.drop(['pet_owners_cell_number', 'pet_owners_work_phone', 'pet_owners_home_number', 'current_foster_adopter', 'previous_owner_information'], axis = 1)

In [ ]:
# Change all text in dataframe to lowercase
df = df.apply(lambda x: x.astype(str).str.lower())

In [ ]:
# Check pet_name type
print("Old type: ", df.pet_name.dtypes)
# Change pet_name entries to strings
df.pet_name = df.pet_name.astype('string')
# Check pet_name type again
print("New type: ", df.pet_name.dtypes)
# Change the strings to lowercase
df.pet_name = df.pet_name.str.lower()

Old type:  object
New type:  string


In [ ]:
df.head(1)

pet_name            pet_breed  \
0  7 kittens litter: henry (courtesy post)- no lo...  american short hair   

  date_pet_entered_your_care length_of_stay_days adoption_foster_date  \
0                  6/18/2022                  18                  nan   

  last_status_change     current_status pet_type pet_age petstablished_id  \
0          21-nov-22  returned to owner      cat  kitten        ps1450793   

     size weight         color shedding coat_length temperament breed_type  \
0  medium    nan  black, white      nan         nan         nan        nan   

  date_placed_in_current_location microchip_id microchip_manufacturer  \
0                             nan          nan                    nan   

  adoption_fee internal_id animal_control_id adoption_application_visible  \
0           75         nan               nan                    invisible   

  foster_application_visible last_updated_at gender acquired_by date_of_birth  \
0                  invisible       21-nov-22   male         nan     4/20/2022   

        age_in_years declawed_status coat_pattern events_attendance is_mix  \
0  1 years, 8 months             nan          nan               nan    yes   

  shots_up_to_date spayed_neutered hypoallergenic housebroken special_need  \
0              yes        not sure       not sure         yes     not sure   

  needs_foster gets_along_with_cats gets_along_with_dogs gets_along_with_kids  \
0           no             not sure             not sure             not sure   

                                         description  \
0  <p>henry is one of seven kitty siblings (b. 4/...   

                                      internal_notes  \
0  reason rehoming:fostering them until finding h...   

  behavioral_tracking_notes has_finalized_application  \
0                       nan                        no   

  where_was_pet_originally_found date_pet_entered_your_care_1  \
0                            nan                    6/18/2022   

  rabies_issue_date additional_comments owner_requested_euthanasia  \
0               nan                 nan                        nan   

  additional_internal_comments type_of_intake parents  \
0                          nan          other     nan   

                                            siblings euthanasia_date  \
0  7 kittens litter: pumpkin (courtesy post)  ps1...             nan   

  euthanasia_reason date_transferred transferred_from transferred_to  \
0               nan              nan              nan            nan   

  transfer_reason deceased_date deceased_reason cross_post_organization_name  \
0             nan           nan             nan                          nan   

  expected_euthanasia_date transferred_out_location transferred_in  
0                      nan                      nan            nan

In [ ]:
# Drop all columns that will not be used in the analysis
df = df.drop(columns = ['date_pet_entered_your_care', 'adoption_foster_date', 'last_status_change', 'weight', 'color', 'date_placed_in_current_location', 
                        'microchip_id', 'microchip_manufacturer', 'internal_id', 'animal_control_id', 'adoption_application_visible', 'foster_application_visible', 
                        'last_updated_at', 'acquired_by', 'date_of_birth', 'declawed_status', 'coat_pattern', 'events_attendance', 'needs_foster', 'description', 
                        'internal_notes', 'behavioral_tracking_notes', 'has_finalized_application', 'where_was_pet_originally_found', 'date_pet_entered_your_care_1', 
                        'rabies_issue_date', 'additional_comments', 'owner_requested_euthanasia', 'additional_internal_comments', 'type_of_intake', 'siblings', 
                        'euthanasia_date', 'euthanasia_reason', 'date_transferred', 'transferred_from', 'transferred_to', 'transfer_reason', 'deceased_date', 
                        'deceased_reason', 'cross_post_organization_name', 'expected_euthanasia_date', 'transferred_out_location', 'transferred_in', 'parents'], axis = 1)

In [ ]:
df.head(1)

pet_name            pet_breed  \
0  7 kittens litter: henry (courtesy post)- no lo...  american short hair   

  length_of_stay_days     current_status pet_type pet_age petstablished_id  \
0                  18  returned to owner      cat  kitten        ps1450793   

     size shedding coat_length temperament breed_type adoption_fee gender  \
0  medium      nan         nan         nan        nan           75   male   

        age_in_years is_mix shots_up_to_date spayed_neutered hypoallergenic  \
0  1 years, 8 months    yes              yes        not sure       not sure   

  housebroken special_need gets_along_with_cats gets_along_with_dogs  \
0         yes     not sure             not sure             not sure   

  gets_along_with_kids  
0             not sure

In [ ]:
len(df)

3776

In [ ]:
df.length_of_stay_days.dtypes

dtype('O')

In [ ]:
df.length_of_stay_days =  df.length_of_stay_days.astype(int)

In [ ]:
df.length_of_stay_days.max()

2458014

In [ ]:
# Drop row where length of stay is greater than 2000
df = df[df.length_of_stay_days < 2000]

In [ ]:
df.length_of_stay_days.max()

1807

In [ ]:
# Remove all animals besides dogs from analysis
df = df[df['pet_type'] == 'dog']

In [ ]:
len(df)

3188

With all animals, the number of records was 3776. After removing animals that were not dogs, there were 3193 records.

In [ ]:
# Ascertain whether the pet_name column contains 'Courtesy' and make a new column called 'courtesy_post' with yes or no
df['courtesy_post'] = df.pet_name.apply(lambda row: 'yes' if 'courtesy' in row else 'no')

In [ ]:
# Ascertain whether the pet_name column contains 'cross post' and create a new column called 'cross_post' with yes or no
df['cross_post'] = df.pet_name.apply(lambda row: 'yes' if 'cross' in row else 'no')

In [ ]:
df.head(1)

pet_name                                   pet_breed  length_of_stay_days  \
5    abbey  schnauzer (standard), jack russell terrier                   43   

  current_status pet_type pet_age petstablished_id   size     shedding  \
5        adopted      dog  senior        ps1120886  small  no shedding   

  coat_length temperament breed_type adoption_fee  gender        age_in_years  \
5         nan         nan        nan          400  female  13 years, 6 months   

  is_mix shots_up_to_date spayed_neutered hypoallergenic housebroken  \
5    yes              yes             yes            yes         yes   

  special_need gets_along_with_cats gets_along_with_dogs gets_along_with_kids  \
5           no                  yes                  yes                  yes   

  courtesy_post cross_post  
5            no         no

In [ ]:
# Add a new column called 'litter'
df['litter'] = ''

In [ ]:
# Initiate a list that contains the strings that indicate an animal is part of a litter
litter_list = ["'s", 'tx', 'ok', 'oklahoma', 'litter', 'kittens', 'kitten', 'pup', 'puppies', 'kitten', 'precious gems']

In [ ]:
# Put 'yes' in the litter column if strings from litter_list are in the pet_name column, otherwise put 'no'
df['litter'] = df['pet_name'].apply(lambda x: 'yes' if any(s in x for s in litter_list) else 'no')

In [ ]:
# Check the unique values in the litter column to verify that the previous function was run correctly
df.litter.unique()

array(['no', 'yes'], dtype=object)

In [ ]:
# Add a new column called 'litter_parent'
df['parent_of_litter'] = ''

In [ ]:
# Initiate a list including common ways to refer to parents
litter_parent_list = ['mama', 'mom', 'mother', 'dad', 'papa', 'pa', 'ma', 'father']

In [ ]:
# Put 'yes' in the parent_of_litter column if strings from litter_parent_list are in the pet_name column, otherwise put 'no'
df['parent_of_litter'] = df['parent_of_litter'].apply(lambda x: 'yes' if any(s in x for s in litter_parent_list) else 'no')

In [ ]:
# Drop the pet_name column now that useful information has been extracted
df = df.drop(['pet_name'], axis = 1)

In [ ]:
df.head()

pet_breed  length_of_stay_days  \
5   schnauzer (standard), jack russell terrier                   43   
6                         old english sheepdog                    2   
7                        schnauzer (miniature)                   24   
9          westie, west highland white terrier                   35   
10                                   chihuahua                  113   

   current_status pet_type pet_age petstablished_id   size     shedding  \
5         adopted      dog  senior        ps1120886  small  no shedding   
6        deceased      dog   puppy         ps945343  small          nan   
7         adopted      dog  senior         ps333886  small          nan   
9         adopted      dog   adult        ps1251360  small          nan   
10       deceased      dog   adult         ps337595  small          nan   

   coat_length temperament breed_type adoption_fee  gender  \
5          nan         nan        nan          400  female   
6          nan         nan        nan          nan  female   
7          nan         nan        nan          250  female   
9          nan         nan        nan          365  female   
10         nan         nan        nan          100  female   

          age_in_years is_mix shots_up_to_date spayed_neutered hypoallergenic  \
5   13 years, 6 months    yes              yes             yes            yes   
6    3 years, 1 months     no         not sure        not sure       not sure   
7   14 years, 2 months     no              yes             yes            yes   
9   10 years, 1 months     no              yes             yes            yes   
10  12 years, 2 months     no              yes             yes             no   

   housebroken special_need gets_along_with_cats gets_along_with_dogs  \
5          yes           no                  yes                  yes   
6     not sure     not sure             not sure             not sure   
7          yes           no             not sure                  yes   
9          yes     not sure             not sure             not sure   
10         yes          yes             not sure                  yes   

   gets_along_with_kids courtesy_post cross_post litter parent_of_litter  
5                   yes            no         no     no               no  
6              not sure            no         no     no               no  
7              not sure            no         no     no               no  
9                    no            no         no     no               no  
10                  yes            no         no     no               no

In [ ]:
# Check the first few entries for pet_breed
df.pet_breed.head(4)

5    schnauzer (standard), jack russell terrier
6                          old english sheepdog
7                         schnauzer (miniature)
9           westie, west highland white terrier
Name: pet_breed, dtype: object

In [ ]:
# Change pet_breed from string to list (using commas as separators)
def to_list(value):
    return value.split(',')

df.pet_breed = df.pet_breed.apply(to_list)

In [ ]:
# Check the first few entries for pet_breed, they should now be lists
df.pet_breed.head(4)

5    [schnauzer (standard),  jack russell terrier]
6                           [old english sheepdog]
7                          [schnauzer (miniature)]
9           [westie,  west highland white terrier]
Name: pet_breed, dtype: object

In [ ]:
# Create a data frame with only the breeds
breed_df = df.pet_breed.apply(pd.Series)
# Combine the breed data frame with the original data frame
df = pd.concat([df, breed_df], axis = 1)
# Drop the pet_breed column
df = df.drop('pet_breed', axis = 1)

In [ ]:
# Rename the breed columns
df = df.rename(columns = {0:'primary_breed', 1:'secondary_breed', 2:'third_breed', 3:'fourth_breed', 4:'fifth_breed'})

In [ ]:
df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
5                    43        adopted      dog  senior        ps1120886   
6                     2       deceased      dog   puppy         ps945343   
7                    24        adopted      dog  senior         ps333886   
9                    35        adopted      dog   adult        ps1251360   
10                  113       deceased      dog   adult         ps337595   

     size     shedding coat_length temperament breed_type adoption_fee  \
5   small  no shedding         nan         nan        nan          400   
6   small          nan         nan         nan        nan          nan   
7   small          nan         nan         nan        nan          250   
9   small          nan         nan         nan        nan          365   
10  small          nan         nan         nan        nan          100   

    gender        age_in_years is_mix shots_up_to_date spayed_neutered  \
5   female  13 years, 6 months    yes              yes             yes   
6   female   3 years, 1 months     no         not sure        not sure   
7   female  14 years, 2 months     no              yes             yes   
9   female  10 years, 1 months     no              yes             yes   
10  female  12 years, 2 months     no              yes             yes   

   hypoallergenic housebroken special_need gets_along_with_cats  \
5             yes         yes           no                  yes   
6        not sure    not sure     not sure             not sure   
7             yes         yes           no             not sure   
9             yes         yes     not sure             not sure   
10             no         yes          yes             not sure   

   gets_along_with_dogs gets_along_with_kids courtesy_post cross_post litter  \
5                   yes                  yes            no         no     no   
6              not sure             not sure            no         no     no   
7                   yes             not sure            no         no     no   
9              not sure                   no            no         no     no   
10                  yes                  yes            no         no     no   

   parent_of_litter          primary_breed               secondary_breed  \
5                no   schnauzer (standard)          jack russell terrier   
6                no   old english sheepdog                           NaN   
7                no  schnauzer (miniature)                           NaN   
9                no                 westie   west highland white terrier   
10               no              chihuahua                           NaN   

   third_breed fourth_breed fifth_breed  
5          NaN          NaN         NaN  
6          NaN          NaN         NaN  
7          NaN          NaN         NaN  
9          NaN          NaN         NaN  
10         NaN          NaN         NaN

In [ ]:
df = df.applymap(lambda x: x.lstrip() if isinstance(x, str) else x)

C:\Users\runek\AppData\Local\Temp\ipykernel_7352\1050543379.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: x.lstrip() if isinstance(x, str) else x)


In [ ]:
len(df)

3188

In [ ]:
# Remove dogs for which we don't know the breed
exclude_breed = 'unknown'
df = df[~df.primary_breed.str.contains(exclude_breed, case = False)]

In [ ]:
len(df)

3127

In [ ]:
# Strip leading spaces from the data frame
df = df.applymap(lambda x: x.lstrip() if isinstance(x, str) else x)

C:\Users\runek\AppData\Local\Temp\ipykernel_7352\2185164362.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: x.lstrip() if isinstance(x, str) else x)


In [ ]:
breed_list = [
    'akita',
    'alaskan malamute',
    'american bulldog',
    'american eskimo dog',
    'american pit bull terrier',
    'american staffordshire terrier',
    'anatolian shepherd',
    'australian cattle dog',
    'australian shepherd',
    'australian terrier',
    'basenji',
    'belgian malinois',
    'belgian shepherd',
    'bernese mountain dog',
    'bichon frise',
    'black and tan coonhound',
    'black mouth cur',
    'bloodhound',
    'blue heeler',
    'bluetick coonhound',
    'border collie',
    'boston terrier',
    'boxer',
    'brittany',
    'bulldog',
    'bullmastiff',
    'bull terrier',
    'cane corso',
    'carolina dog',
    'catahoula leopard dog',
    'chesapeake bay retriever',
    'chihuahua',
    'chiweenie',
    'chinese crested',
    'chow chow',
    'cocker spaniel',
    'collie',
    'corgi',
    'coton de tulear',
    'dachshund',
    'dalmatian',
    'doberman pinscher',
    'dutch shepherd',
    'english bulldog',
    'english mastiff',
    'english pointer',
    'english redtick coonhound',
    'english setter',
    'english shepherd',
    'french bulldog',
    'german shorthaired pointer',
    'german wirehaired pointer',
    'german shepherd dog', 
    'golden retriever',
    'goldendoodle',
    'greyhound',
    'havanese',
    'irish setter',
    'irish wolfhound',
    'jack russell terrier',
    'japanese chin',
    'labradoodle',
    'lancashire heeler',
    'lhasa apso',
    'maltese',
    'mastiff',
    'miniature dachshund',
    'miniature pinscher',
    'miniature poodle',
    'mixed breed',
    '31 flavors mixes',
    'mountain cur',
    'neapolitan mastiff',
    'newfoundland',
    'norwegian elkhound',
    'old english sheepdog',
    'other',
    'papillon',
    'patterdale terrier (fell terrier)',
    'pekingese',
    'pit bull terrier', 
    'plott hound',
    'pointer',
    'pomeranian',
    'poodle (miniature)',
    'poodle (standard)',
    'poodle (toy or teacup)',
    'poodles (all types)',
    'preso canario',
    'rat terrier',
    'redbone coonhound',
    'rhodesian ridgeback',
    'saint bernard',
    'schipperke',
    'shar pei',
    'sheltie',
    'shiba inu',
    'silky terrier',
    'snow dogs (husky/fluffy types)',
    'schnauzer (standard)',
    'schnauzer (miniature)',
    'schnauzer (giant)',
    'setter (unknown type)',
    'shepherds (all types)',
    'shepherd (unknown type)',
    'shih tzu',
    'siberian husky',
    'silky terrier',
    'springer spaniel',
    'st. bernard',
    'terrier (unknown type',
    'terrier (medium)',
    'terrier (small)',
    'toy poodle',
    'treeing walker coonhound',
    # 'unknown',
    'vizsla',
    'weimaraner',
    'welsh corgi',
    'welsh terrier',
    'whippet',
    'wheaten terrier',
    'wirehaired fox terrier',
    'wirehaired pointing griffon',
    'xoloitzcuintle/mexican hairless'
]

In [ ]:
# Remove all the names of breeds that aren't in the breed list (for example, remove 'westie')
df.primary_breed = df.primary_breed.where(df.primary_breed.isin(breed_list), None)

If there are null entries in a breed column, grab the entry from the column next to it. Continue until every primary breed row is filled out, then shift until there are as few nulls as possible.

In [ ]:
for index, row in df.iterrows():
    if row.primary_breed is None:
        df.loc[index, 'primary_breed'] = row['secondary_breed']
        df.loc[index, 'secondary_breed'] = None

In [ ]:
for index, row in df.iterrows():
    if row.secondary_breed is None:
        df.loc[index, 'secondary_breed'] = row['third_breed']
        df.loc[index, 'third_breed'] = None

In [ ]:
for index, row in df.iterrows():
    if row.third_breed is None:
        df.loc[index, 'third_breed'] = row['fourth_breed']
        df.loc[index, 'fourth_breed'] = None

In [ ]:
for index, row in df.iterrows():
    if row.fourth_breed is None:
        df.loc[index, 'third_breed'] = row['fourth_breed']
        df.loc[index, 'fourth_breed'] = None

In [ ]:
df.fifth_breed = df.fifth_breed.where(df.fifth_breed.isin(breed_list), None)

In [ ]:
df = df.drop(columns = ['fifth_breed'], axis = 1)

In [ ]:
df.fourth_breed = df.fourth_breed.where(df.fourth_breed.isin(breed_list), None)
df.third_breed = df.third_breed.where(df.third_breed.isin(breed_list), None)

In [ ]:
pd.set_option('display.max_rows', None)
df.primary_breed.value_counts()

primary_breed
chihuahua                            237
pit bull terrier                     214
german shepherd dog                  206
boxer                                168
yorkshire terrier                    131
shih tzu                             110
australian shepherd                   91
maltese                               81
border collie                         71
shepherds (all types)                 64
31 flavors mixes                      53
pomeranian                            51
dachshund                             48
blue heeler                           41
australian cattle dog                 34
jack russell terrier                  33
poodle (miniature)                    26
rat terrier                           26
miniature pinscher                    26
golden retriever                      25
goldendoodle                          24
bichon frise                          24
cocker spaniel                        24
corgi                                 23
ch

In [ ]:
pd.reset_option('display.max_rows')

In [ ]:
# Define a list of bully breeds (these breeds have a bad reputation)
bully_breeds = [
    'american bulldog',
    'american pit bull terrier',
    'american staffordshire terrier',
    'boston terrier',
    'bullmastiff',
    'bull terrier',
    'cane corso',
    'carolina dog',
    # 'french bulldog',
    'staffordshire bull terrier'
]

In [ ]:
# Create a column 'bully_breed' and put 'yes' if the breed columns contain a string from bully_breeds, otherwise put 'no'
df['bully_breed'] = df[['primary_breed', 'secondary_breed', 'third_breed', 'fourth_breed']].isin(bully_breeds).any(axis = 1)
df['bully_breed'] = df['bully_breed'].map({True:'yes', False:'no'})

In [ ]:
df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
5                    43        adopted      dog  senior        ps1120886   
6                     2       deceased      dog   puppy         ps945343   
7                    24        adopted      dog  senior         ps333886   
9                    35        adopted      dog   adult        ps1251360   
10                  113       deceased      dog   adult         ps337595   

     size     shedding coat_length temperament breed_type adoption_fee  \
5   small  no shedding         nan         nan        nan          400   
6   small          nan         nan         nan        nan          nan   
7   small          nan         nan         nan        nan          250   
9   small          nan         nan         nan        nan          365   
10  small          nan         nan         nan        nan          100   

    gender        age_in_years is_mix shots_up_to_date spayed_neutered  \
5   female  13 years, 6 months    yes              yes             yes   
6   female   3 years, 1 months     no         not sure        not sure   
7   female  14 years, 2 months     no              yes             yes   
9   female  10 years, 1 months     no              yes             yes   
10  female  12 years, 2 months     no              yes             yes   

   hypoallergenic housebroken special_need gets_along_with_cats  \
5             yes         yes           no                  yes   
6        not sure    not sure     not sure             not sure   
7             yes         yes           no             not sure   
9             yes         yes     not sure             not sure   
10             no         yes          yes             not sure   

   gets_along_with_dogs gets_along_with_kids courtesy_post cross_post litter  \
5                   yes                  yes            no         no     no   
6              not sure             not sure            no         no     no   
7                   yes             not sure            no         no     no   
9              not sure                   no            no         no     no   
10                  yes                  yes            no         no     no   

   parent_of_litter                primary_breed       secondary_breed  \
5                no         schnauzer (standard)  jack russell terrier   
6                no         old english sheepdog                   NaN   
7                no        schnauzer (miniature)                   NaN   
9                no  west highland white terrier                   NaN   
10               no                    chihuahua                   NaN   

   third_breed fourth_breed bully_breed  
5         None         None          no  
6         None         None          no  
7         None         None          no  
9         None         None          no  
10        None         None          no

In [ ]:
# Drop dogs with primary breeds that have fewer than 10 entries
min_number = 10
value_counts = df.primary_breed.value_counts()

values_to_drop = value_counts[value_counts < min_number].index

df = df[~df.primary_breed.isin(values_to_drop)]

In [ ]:
len(df)

2868

In [ ]:
# Drop rows where primary_breed is null
df = df.dropna(subset = ['primary_breed'])

In [ ]:
len(df)

2374

In [ ]:
# Create a list of the most popular dog breeds in the US in 2022
popular_breeds_2022 = ['labrador retriever', 'german shepherd', 'golden retriever', 'french bulldog', 'bulldog', 'poodle', 'beagle', 'rottweiler', 'yorkshire terrier', 'dachshund']

In [ ]:
# Create a column 'popular_breed' that has 'yes' if the dog is a popular breed, otherwise put 'no'
df['popular_breed'] = df[['primary_breed', 'secondary_breed', 'third_breed', 'fourth_breed']].isin(popular_breeds_2022).any(axis = 1)
df['popular_breed'] = df['popular_breed'].map({True:'yes', False:'no'})

Edit the adoption fee column and create a new column 'bonded_pair' that says yes if the dog is part of a pair, and 'no' if not

In [ ]:
# Add a new column for bonded pairs. This will allow me to keep potentially valuable information.
df['bonded_pair'] = ''

In [ ]:
# Show the current type of the adoption fee column
print('Old adoption fee type: ', df.adoption_fee.dtypes)
# Change the adoption fee column type to string (this will make it easier to work with later)
df.adoption_fee = df.adoption_fee.astype('string')
# Show the new type of the adoption fee column
print('New adoption fee type: ', df.adoption_fee.dtypes)

Old adoption fee type:  object
New adoption fee type:  string


In [ ]:
# Show the current type of the bonded pair column
print('Old bonded pair type: ', df.bonded_pair.dtypes)
# Change the bonded pair column type to string
df.bonded_pair = df.bonded_pair.astype('string')
# Show the new type of the bonded pair column
print('New bonded pair type: ', df.bonded_pair.dtypes)

Old bonded pair type:  object
New bonded pair type:  string


In [ ]:
# Fill the NaN values with an empty string - this will prevent errors with iterating through the rows
df.adoption_fee = df.adoption_fee.fillna('')

In [ ]:
# If the adoption fee column contains an indication that an animal is part of a bonded pair, paste the value of adoption fee
# into the bonded pair column
df.bonded_pair = df.adoption_fee.apply(lambda row: row[0:row.find('for|both|and|together|togethr')] if ' ' in row else None)

In [ ]:
# Check that the values of adoption fee have been transferred over to bonded pair
df.bonded_pair.iloc[60:70]

100           None
101           None
103           None
106           None
108           None
109           None
113    730 for bot
114           None
115           None
117           None
Name: bonded_pair, dtype: object

In [ ]:
# Replace NaN values in the bonded pair column with empty strings (prevents iteration errors later)
df.bonded_pair = df.bonded_pair.fillna('')

In [ ]:
# Change the bonded_pair column rows to 'yes' if they contain strings that indicate bonded pair
df.loc[df.bonded_pair.str.contains('for|both|and|together|togethr'), 'bonded_pair'] = 'yes'

In [ ]:
# Check to see if the strings have been changed
df.bonded_pair.iloc[60:70]

100       
101       
103       
106       
108       
109       
113    yes
114       
115       
117       
Name: bonded_pair, dtype: object

In [ ]:
# Replace the empty strings with the word 'no' to indicate a dog is NOT part of a bonded pair
df.bonded_pair = df.bonded_pair.replace('', 'no')

In [ ]:
# Check to see if the strings have been changed
df.bonded_pair.iloc[60:70]

100     no
101     no
103     no
106     no
108     no
109     no
113    yes
114     no
115     no
117     no
Name: bonded_pair, dtype: object

In [ ]:
# Replace NaN values in the adoption fee column with empty strings (prevents iteration errors later)
df.adoption_fee = df.adoption_fee.fillna('')

In [ ]:
# Split the row values into a list on spaces if there are spaces in the row (this captures any of the bonded pair indications)
df.adoption_fee = df.adoption_fee.apply(lambda row: row.split(' ') if ' ' in row else row)
# If the row type is a list (it had spaces and was changed above), keep only the first value in the list
df.adoption_fee = df.adoption_fee.apply(lambda row: row.pop(0) if type(row) == list else row)

In [ ]:
# Check that the bonded pair strings have been removed
df.adoption_fee.iloc[60:70]

100    450
101    400
103    450
106    515
108    365
109    265
113    730
114    250
115    425
117    365
Name: adoption_fee, dtype: object

In [ ]:
# There is at least one entry that contains something like '250-275'. I want to get rid of that and keep only one value
# Split the row values into a list on - if there is a - in the row
df.adoption_fee = df.adoption_fee.apply(lambda row: row.split('-') if '-' in row else row)
# If the row type is a list (it had - and was changed above), keep only the first value in the list
df.adoption_fee = df.adoption_fee.apply(lambda row: row.pop(0) if type(row) == list else row)

In [ ]:
# There is an entry called 'please contact muddy paws...'. I want to get rid of that.
# Split the rows into a list on 'please' if they contain 'please'
df.adoption_fee = df.adoption_fee.apply(lambda row: row.split('please') if 'please' in row else row)
# If the row type is a list (it had 'please' and was changed avove), keep only the first value in the list (which should be blank)
df.adoption_fee = df.adoption_fee.apply(lambda row: row.pop(0) if type(row) == list else row)

In [ ]:
# Check the adoption fee column again
df.adoption_fee.tail(10)

3763     400
3764     350
3765     400
3766    $400
3767     350
3769     465
3771     150
3773     300
3774     400
3775     415
Name: adoption_fee, dtype: object

In [ ]:
# Drop $ from the column
df.adoption_fee = df.adoption_fee.str.replace('\$', '', regex = True)

In [ ]:
df.adoption_fee.tail(10)

3763    400
3764    350
3765    400
3766    400
3767    350
3769    465
3771    150
3773    300
3774    400
3775    415
Name: adoption_fee, dtype: object

Clean the age column, and add a new column that contains only age in months.

In [ ]:
# Check the age in years column
df.age_in_years

5       13 years, 6 months
7       14 years, 2 months
10      12 years, 2 months
11      16 years, 7 months
14       5 years, 5 months
               ...        
3769     2 years, 8 months
3771                   nan
3773    10 years, 2 months
3774    3 years, 11 months
3775     2 years, 8 months
Name: age_in_years, Length: 2374, dtype: object

In [ ]:
# Add an column for the months values in the age in years column to be split into
df.insert(33, 'age_in_months', '')

# Show the current type of the age_in_years column
print('Old age in years type: ', df.age_in_years.dtypes)
# Change the type of the age_in_years column to string
df.age_in_years = df.age_in_years.astype('string')
# Show the new type of the age_in_years column
print('New age in years type: ', df.age_in_years.dtypes)

Old age in years type:  object
New age in years type:  string


In [ ]:
# Fill the NaN values with blank strings (they can't be iterated through)
df.age_in_years = df.age_in_years.fillna('')

# If the word 'months' is in the age_in_years column, paste the whole value of age_in_years column to the new age_in_months column
df.age_in_months = df.age_in_years.apply(lambda row: row[0:row.find('months')] if 'months' in row else None)


In [ ]:
df.age_in_months

5       13 years, 6 
7       14 years, 2 
10      12 years, 2 
11      16 years, 7 
14       5 years, 5 
            ...     
3769     2 years, 8 
3771            None
3773    10 years, 2 
3774    3 years, 11 
3775     2 years, 8 
Name: age_in_months, Length: 2374, dtype: object

In [ ]:
# I want to remove the months portion from the age_in_years column
# Split age_in_years values to a list on the , if they have a ,
df.age_in_years = df.age_in_years.apply(lambda row: row.split(',') if ',' in row else row)
# Keep only the first value of the list if the row type is a list (changed above)
df.age_in_years = df.age_in_years.apply(lambda row: row.pop(0) if type(row) == list else row)

In [ ]:
# Check the age_in_years column to ensure that months have been removed
df.age_in_years

5       13 years
7       14 years
10      12 years
11      16 years
14       5 years
          ...   
3769     2 years
3771         nan
3773    10 years
3774     3 years
3775     2 years
Name: age_in_years, Length: 2374, dtype: object

In [ ]:
# Check the age_in_months column - it shows that years have not yet been removed
df.age_in_months

5       13 years, 6 
7       14 years, 2 
10      12 years, 2 
11      16 years, 7 
14       5 years, 5 
            ...     
3769     2 years, 8 
3771            None
3773    10 years, 2 
3774    3 years, 11 
3775     2 years, 8 
Name: age_in_months, Length: 2374, dtype: object

In [ ]:
# Fill the NaN values in age_in_months with an empty string (makes iterating possible)
df.age_in_months = df.age_in_months.fillna('')

In [ ]:
# Split the age_in_months values into lists on the , if they contain ,
df.age_in_months = df.age_in_months.apply(lambda row: row.split(',') if ',' in row else row)
# Keep only the last value of the lists if the row type is a list (changed above)
df.age_in_months = df.age_in_months.apply(lambda row: row.pop(-1) if type(row) == list else row)

In [ ]:
# Check age_in_months to be sure that the years values were removed
df.age_in_months

5         6 
7         2 
10        2 
11        7 
14        5 
        ... 
3769      8 
3771        
3773      2 
3774     11 
3775      8 
Name: age_in_months, Length: 2374, dtype: object

In [ ]:
# Review the age_in_years column
df.age_in_years

5       13 years
7       14 years
10      12 years
11      16 years
14       5 years
          ...   
3769     2 years
3771         nan
3773    10 years
3774     3 years
3775     2 years
Name: age_in_years, Length: 2374, dtype: object

In [ ]:
# We can see above that there is a strange value (-283 days). I want to remove that value completely
# If there is a - in the row, split into a list on the -
df.age_in_years = df.age_in_years.apply(lambda row: row.split('-') if '-' in row else row)
# If the type of the row is a list (changed above) keep only the first value of the list (which in this case should be blank)
df.age_in_years = df.age_in_years.apply(lambda row: row.pop(0) if type(row) == list else row)

In [ ]:
# Check to be sure the negative value was removed
df.age_in_years

5       13 years
7       14 years
10      12 years
11      16 years
14       5 years
          ...   
3769     2 years
3771         nan
3773    10 years
3774     3 years
3775     2 years
Name: age_in_years, Length: 2374, dtype: object

In [ ]:
# Remove the 'years' string from the age_in_years column
# Split into lists on the spaces if 'years' is in the row
df.age_in_years = df.age_in_years.apply(lambda row: row.split(' ') if 'years' in row else row)
# Keep only the first value of the list if the row type is a list (changed above)
df.age_in_years = df.age_in_years.apply(lambda row: row.pop(0) if type(row) == list else row)

In [ ]:
# Check age_in_years to be sure 'years' was removed
df.age_in_years

5        13
7        14
10       12
11       16
14        5
       ... 
3769      2
3771    nan
3773     10
3774      3
3775      2
Name: age_in_years, Length: 2374, dtype: object

In [ ]:
# Add a new column titled 'age_years_to_months'
# I will be using this column to calculate how many months the years in the age_in_years column are equal to
df.insert(34, 'age_years_to_months', '')

In [ ]:
# Show the current type of the age_in_years column
print('Old age in years type: ', df.age_in_years.dtypes)
# Change the type of the age_in_years column to numerical values, coercing errors
df.age_in_years = pd.to_numeric(df.age_in_years, errors = 'coerce')
# Change the type of the age_in_years column to float values
df.age_in_years = df.age_in_years.astype('float')
# Sho the new type of the age_in_years column
print('New age in years type: ', df.age_in_years.dtypes)

Old age in years type:  object
New age in years type:  float64


In [ ]:
# Check the age_in_years column
df.age_in_years

5       13.0
7       14.0
10      12.0
11      16.0
14       5.0
        ... 
3769     2.0
3771     NaN
3773    10.0
3774     3.0
3775     2.0
Name: age_in_years, Length: 2374, dtype: float64

In [ ]:
# Set the age_years_to_months column to be equal to the age_in_years column * 12
df.age_years_to_months = df.age_in_years * 12

In [ ]:
# Check the age_years_to_months column to verify that it has been changed
df.age_years_to_months

5       156.0
7       168.0
10      144.0
11      192.0
14       60.0
        ...  
3769     24.0
3771      NaN
3773    120.0
3774     36.0
3775     24.0
Name: age_years_to_months, Length: 2374, dtype: float64

In [ ]:
# Show the current type of the age_in_months column
print('Old age in months type: ', df.age_in_months.dtypes)
# Change the type of the age_in_months column, coercing errors
df.age_in_months = pd.to_numeric(df.age_in_months, errors = 'coerce')
# Change the type of the age_in_months column to float
df.age_in_months = df.age_in_months.astype('float')
# Show the new type of the age_in_months column
print('New age in months type: ', df.age_in_months.dtypes)

Old age in months type:  object
New age in months type:  float64


In [ ]:
# Add a new column called age_total_months to reflect the actual age of the dog in months
df['age_total_months'] = ''

# Set the age_total_months column to be equal to age_years_to_months + age_in_months
df.age_total_months = df.age_years_to_months + df.age_in_months

In [ ]:
# Check that the values in age_total_months make sense
df.age_total_months

5       162.0
7       170.0
10      146.0
11      199.0
14       65.0
        ...  
3769     32.0
3771      NaN
3773    122.0
3774     47.0
3775     32.0
Name: age_total_months, Length: 2374, dtype: float64

In [ ]:
df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
5                    43        adopted      dog  senior        ps1120886   
7                    24        adopted      dog  senior         ps333886   
10                  113       deceased      dog   adult         ps337595   
11                   99        adopted      dog  senior          ps35514   
14                   31        adopted      dog   adult        ps1564518   

     size        shedding coat_length temperament breed_type adoption_fee  \
5   small     no shedding         nan         nan        nan          400   
7   small             nan         nan         nan        nan          250   
10  small             nan         nan         nan        nan          100   
11  small  sheds a little       short         nan  purebreed          200   
14  small     no shedding         nan         nan        nan          550   

    gender  age_in_years is_mix shots_up_to_date spayed_neutered  \
5   female          13.0    yes              yes             yes   
7   female          14.0     no              yes             yes   
10  female          12.0     no              yes             yes   
11  female          16.0     no              yes             yes   
14    male           5.0     no              yes             yes   

   hypoallergenic housebroken special_need gets_along_with_cats  \
5             yes         yes           no                  yes   
7             yes         yes           no             not sure   
10             no         yes          yes             not sure   
11             no          no           no             not sure   
14       not sure          no     not sure             not sure   

   gets_along_with_dogs gets_along_with_kids courtesy_post cross_post litter  \
5                   yes                  yes            no         no     no   
7                   yes             not sure            no         no     no   
10                  yes                  yes            no         no     no   
11                  yes                  yes           yes         no     no   
14                  yes             not sure            no         no     no   

   parent_of_litter          primary_breed       secondary_breed third_breed  \
5                no   schnauzer (standard)  jack russell terrier        None   
7                no  schnauzer (miniature)                   NaN        None   
10               no              chihuahua                   NaN        None   
11               no              dachshund                   NaN        None   
14               no                maltese                   NaN        None   

   fourth_breed bully_breed popular_breed bonded_pair  age_in_months  \
5          None          no            no          no            6.0   
7          None          no            no          no            2.0   
10         None          no            no          no            2.0   
11         None          no           yes          no            7.0   
14         None          no            no          no            5.0   

    age_years_to_months  age_total_months  
5                 156.0             162.0  
7                 168.0             170.0  
10                144.0             146.0  
11                192.0             199.0  
14                 60.0              65.0

In [ ]:
df['age_total_years'] = ''

df['age_months_to_years'] = ''

df.age_months_to_years = df.age_in_months / 12

In [ ]:
df.age_months_to_years

5       0.500000
7       0.166667
10      0.166667
11      0.583333
14      0.416667
          ...   
3769    0.666667
3771         NaN
3773    0.166667
3774    0.916667
3775    0.666667
Name: age_months_to_years, Length: 2374, dtype: float64

In [ ]:
# Set the age_total_years column to equal age_in_years + age_months_to_years
df.age_total_years = df.age_in_years + df.age_months_to_years

In [ ]:
# Check that the values in age_total_years column make sense
df.age_total_years

5       13.500000
7       14.166667
10      12.166667
11      16.583333
14       5.416667
          ...    
3769     2.666667
3771          NaN
3773    10.166667
3774     3.916667
3775     2.666667
Name: age_total_years, Length: 2374, dtype: float64

In [ ]:
df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
5                    43        adopted      dog  senior        ps1120886   
7                    24        adopted      dog  senior         ps333886   
10                  113       deceased      dog   adult         ps337595   
11                   99        adopted      dog  senior          ps35514   
14                   31        adopted      dog   adult        ps1564518   

     size        shedding coat_length temperament breed_type adoption_fee  \
5   small     no shedding         nan         nan        nan          400   
7   small             nan         nan         nan        nan          250   
10  small             nan         nan         nan        nan          100   
11  small  sheds a little       short         nan  purebreed          200   
14  small     no shedding         nan         nan        nan          550   

    gender  age_in_years is_mix shots_up_to_date spayed_neutered  \
5   female          13.0    yes              yes             yes   
7   female          14.0     no              yes             yes   
10  female          12.0     no              yes             yes   
11  female          16.0     no              yes             yes   
14    male           5.0     no              yes             yes   

   hypoallergenic housebroken special_need gets_along_with_cats  \
5             yes         yes           no                  yes   
7             yes         yes           no             not sure   
10             no         yes          yes             not sure   
11             no          no           no             not sure   
14       not sure          no     not sure             not sure   

   gets_along_with_dogs gets_along_with_kids courtesy_post cross_post litter  \
5                   yes                  yes            no         no     no   
7                   yes             not sure            no         no     no   
10                  yes                  yes            no         no     no   
11                  yes                  yes           yes         no     no   
14                  yes             not sure            no         no     no   

   parent_of_litter          primary_breed       secondary_breed third_breed  \
5                no   schnauzer (standard)  jack russell terrier        None   
7                no  schnauzer (miniature)                   NaN        None   
10               no              chihuahua                   NaN        None   
11               no              dachshund                   NaN        None   
14               no                maltese                   NaN        None   

   fourth_breed bully_breed popular_breed bonded_pair  age_in_months  \
5          None          no            no          no            6.0   
7          None          no            no          no            2.0   
10         None          no            no          no            2.0   
11         None          no           yes          no            7.0   
14         None          no            no          no            5.0   

    age_years_to_months  age_total_months  age_total_years  \
5                 156.0             162.0        13.500000   
7                 168.0             170.0        14.166667   
10                144.0             146.0        12.166667   
11                192.0             199.0        16.583333   
14                 60.0              65.0         5.416667   

    age_months_to_years  
5              0.500000  
7              0.166667  
10             0.166667  
11             0.583333  
14             0.416667

In [ ]:
# Drop the columns that are no longer needed
df = df.drop(['age_in_years', 'age_in_months', 'age_months_to_years', 'age_years_to_months', 'age_total_years'], axis = 1)

In [ ]:
df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
5                    43        adopted      dog  senior        ps1120886   
7                    24        adopted      dog  senior         ps333886   
10                  113       deceased      dog   adult         ps337595   
11                   99        adopted      dog  senior          ps35514   
14                   31        adopted      dog   adult        ps1564518   

     size        shedding coat_length temperament breed_type adoption_fee  \
5   small     no shedding         nan         nan        nan          400   
7   small             nan         nan         nan        nan          250   
10  small             nan         nan         nan        nan          100   
11  small  sheds a little       short         nan  purebreed          200   
14  small     no shedding         nan         nan        nan          550   

    gender is_mix shots_up_to_date spayed_neutered hypoallergenic housebroken  \
5   female    yes              yes             yes            yes         yes   
7   female     no              yes             yes            yes         yes   
10  female     no              yes             yes             no         yes   
11  female     no              yes             yes             no          no   
14    male     no              yes             yes       not sure          no   

   special_need gets_along_with_cats gets_along_with_dogs  \
5            no                  yes                  yes   
7            no             not sure                  yes   
10          yes             not sure                  yes   
11           no             not sure                  yes   
14     not sure             not sure                  yes   

   gets_along_with_kids courtesy_post cross_post litter parent_of_litter  \
5                   yes            no         no     no               no   
7              not sure            no         no     no               no   
10                  yes            no         no     no               no   
11                  yes           yes         no     no               no   
14             not sure            no         no     no               no   

            primary_breed       secondary_breed third_breed fourth_breed  \
5    schnauzer (standard)  jack russell terrier        None         None   
7   schnauzer (miniature)                   NaN        None         None   
10              chihuahua                   NaN        None         None   
11              dachshund                   NaN        None         None   
14                maltese                   NaN        None         None   

   bully_breed popular_breed bonded_pair  age_total_months  
5           no            no          no             162.0  
7           no            no          no             170.0  
10          no            no          no             146.0  
11          no           yes          no             199.0  
14          no            no          no              65.0

In [ ]:
len(df)

2374

In [ ]:
# Remove all animals who are NOT adopted
df = df[df['current_status'] == 'adopted']

In [ ]:
len(df)

1978

Add a sentiment score column based on temperament column

In [ ]:
from textblob import TextBlob

In [ ]:
# Function to get sentiment score
def get_sentiment_score(text):
    if text is not None:
        blob = TextBlob(text)
        return blob.sentiment.polarity
    else:
        return None
    
# Apply the sentiment score function to create a new column
df['sentiment_score'] = df.temperament.apply(get_sentiment_score)

In [ ]:
df.sentiment_score.unique()

array([ 0.        ,  0.3       ,  0.7       ,  0.1       ,  0.455     ,
        0.6       ,  0.5875    , -0.5       ,  0.34166667,  0.375     ,
       -0.24166667, -0.9       ,  0.20714286, -0.25      ,  0.8       ,
        0.91      ,  0.5775    ,  0.35      ,  0.425     , -0.0625    ,
        0.48333333,  0.45      , -0.00666667,  0.25      ])

In [ ]:
df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
5                    43        adopted      dog  senior        ps1120886   
7                    24        adopted      dog  senior         ps333886   
11                   99        adopted      dog  senior          ps35514   
14                   31        adopted      dog   adult        ps1564518   
16                   50        adopted      dog   puppy        ps1357978   

     size        shedding coat_length temperament breed_type adoption_fee  \
5   small     no shedding         nan         nan        nan          400   
7   small             nan         nan         nan        nan          250   
11  small  sheds a little       short         nan  purebreed          200   
14  small     no shedding         nan         nan        nan          550   
16  large             nan         nan         nan        nan          465   

    gender is_mix shots_up_to_date spayed_neutered hypoallergenic housebroken  \
5   female    yes              yes             yes            yes         yes   
7   female     no              yes             yes            yes         yes   
11  female     no              yes             yes             no          no   
14    male     no              yes             yes       not sure          no   
16    male    yes              yes             yes       not sure    not sure   

   special_need gets_along_with_cats gets_along_with_dogs  \
5            no                  yes                  yes   
7            no             not sure                  yes   
11           no             not sure                  yes   
14     not sure             not sure                  yes   
16     not sure             not sure             not sure   

   gets_along_with_kids courtesy_post cross_post litter parent_of_litter  \
5                   yes            no         no     no               no   
7              not sure            no         no     no               no   
11                  yes           yes         no     no               no   
14             not sure            no         no     no               no   
16             not sure            no         no     no               no   

            primary_breed        secondary_breed third_breed fourth_breed  \
5    schnauzer (standard)   jack russell terrier        None         None   
7   schnauzer (miniature)                    NaN        None         None   
11              dachshund                    NaN        None         None   
14                maltese                    NaN        None         None   
16        black mouth cur  shepherds (all types)        None         None   

   bully_breed popular_breed bonded_pair  age_total_months  sentiment_score  
5           no            no          no             162.0              0.0  
7           no            no          no             170.0              0.0  
11          no           yes          no             199.0              0.0  
14          no            no          no              65.0              0.0  
16          no            no          no              25.0              0.0

In [ ]:
# Validate that length of stay amounts make sense
# Check the minimum value in length of stay
print(df.length_of_stay_days.min())
print(df.length_of_stay_days.max())

0
1807


Edit medical data data frame

In [ ]:
med_df.head()

Pet Name      Pet Primary Breed  \
0  7 Kittens Litter: Paisley (Courtesy Post)     American Short Hair   
1    7 Kittens Litter: Piper (Courtesy Post)     American Short Hair   
2  7 Kittens Litter: Pumpkin (Courtesy Post)     American Short Hair   
3                                       Abbey   Schnauzer (Standard)   
4                                        Abby  Schnauzer (Miniature)   

    Pet Secondary Breed Current Pet Status                    Vet Name  \
0                   NaN            Adopted                          .    
1                   NaN            Adopted                          .    
2                   NaN            Adopted  Val Verde Animal Hospital    
3  Jack Russell Terrier            Adopted  Val Verde Animal Hospital    
4                   NaN            Adopted         Ralston Vet Clinic    

          Name  Vet City Vet State Vet Zip Vet Phone Vaccine Name  \
0  Spay/Neuter       NaN       NaN     NaN       NaN          NaN   
1  Spay/Neuter       NaN       NaN     NaN       NaN          NaN   
2  Spay/Neuter  La Vista        NE   68128       NaN          NaN   
3        Other  La Vista        NE   68128       NaN          NaN   
4          NaN   Ralston        NE   68127       NaN          NaN   

   Vaccine Manufacturer Name         Record Type             Subtype  \
0                        NaN  Surgery/Procedures         Spay/Neuter   
1                        NaN  Surgery/Procedures         Spay/Neuter   
2                        NaN  Surgery/Procedures         Spay/Neuter   
3                        NaN  Surgery/Procedures               Other   
4                        NaN  Surgery/Procedures  Dental Prophylaxis   

   Vaccine Serial Number  Vaccine Expiration Date  \
0                    NaN                      NaN   
1                    NaN                      NaN   
2                    NaN                      NaN   
3                    NaN                      NaN   
4                    NaN                      NaN   

                                      Description Follow Up Date  Tag ID  \
0                                             NaN            NaN     NaN   
1                                             NaN            NaN     NaN   
2                                             NaN            NaN     NaN   
3  Growth removal on back leg and dental cleaning            NaN     NaN   
4                                          Dental            NaN     NaN   

  Frequency of Administration Test Result Date of Visit Result  \
0                         NaN         NaN     8/30/2022    NaN   
1                         NaN         NaN     8/30/2022    NaN   
2                         NaN         NaN    11/29/2022    NaN   
3                         NaN         NaN      7/1/2021    NaN   
4                         NaN         NaN     10/1/2018    NaN   

  Expected staples removal date Cost of Visit Petstablished ID Total Cost  \
0                           NaN           NaN        PS1450794     $0.00    
1                           NaN           NaN        PS1450796     $0.00    
2                           NaN           NaN        PS1450791     $0.00    
3                           NaN           NaN        PS1120886     $0.00    
4                           NaN           NaN         PS333886     $0.00    

                                    Description.1  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3  Growth removal on back leg and dental cleaning   
4                                          Dental   

                           Comments  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4  Bloodwork and urinalysis 9/28/18   

                                   Additional Notes  
0                                               NaN  
1            

In [ ]:
# Replace the spaces with _
med_df.columns = med_df.columns.str.replace(' ', '_')
# Change column text to lowercase
med_df.columns = med_df.columns.str.lower()
# Replace ' with empty string
med_df.columns = med_df.columns.str.replace("'", "")
# Replace special characters with empty string
med_df.columns = med_df.columns.str.replace(r'[\?\(\)]', '', regex = True)
# Replace / with _
med_df.columns = med_df.columns.str.replace('/', '_', regex = True)
# Replace .1 with _1
med_df.columns = med_df.columns.str.replace('.1', '_1', regex = True)

In [ ]:
med_df.head(1)

pet_name    pet_primary_breed  \
0  7 Kittens Litter: Paisley (Courtesy Post)   American Short Hair   

  pet_secondary_breed current_pet_status vet_name         name vet_city  \
0                 NaN            Adopted       .   Spay/Neuter      NaN   

  vet_state vet_zip vet_phone vaccine_name  vaccine_manufacturer_name  \
0       NaN     NaN       NaN          NaN                        NaN   

          record_type      subtype  vaccine_serial_number  \
0  Surgery/Procedures  Spay/Neuter                    NaN   

   vaccine_expiration_date description follow_up_date  tag_id  \
0                      NaN         NaN            NaN     NaN   

  frequency_of_administration test_result date_of_visit result  \
0                         NaN         NaN     8/30/2022    NaN   

  expected_staples_removal_date cost_of_visit petstablished_id total_cost  \
0                           NaN           NaN        PS1450794     $0.00    

  description_1 comments additional_notes  
0           NaN      NaN              NaN

In [ ]:
# Drop columns that will not be used in analysis
med_df = med_df.drop(['additional_notes', 'comments', 'description_1', 'total_cost', 'cost_of_visit', 'expected_staples_removal_date', 'result', 'date_of_visit', 
                      'test_result', 'frequency_of_administration', 'tag_id', 'follow_up_date', 'description', 'vaccine_expiration_date', 'vaccine_serial_number', 
                      'vaccine_manufacturer_name', 'vaccine_name', 'vet_phone', 'vet_zip', 'vet_state', 'vet_city', 'name', 'vet_name', 'current_pet_status', 
                      'pet_name', 'pet_primary_breed', 'pet_secondary_breed'], axis = 1)

In [ ]:
med_df.head(1)

record_type      subtype petstablished_id
0  Surgery/Procedures  Spay/Neuter        PS1450794

In [ ]:
len(med_df)

3887

In [ ]:
# Drop all records where the subtype is Spay/Neuter (all their animals get spayed/neutered)
med_df = med_df[med_df['subtype'] != 'Spay/Neuter']

In [ ]:
len(med_df)

1731

In [ ]:
med_df.head(1)

record_type subtype petstablished_id
3  Surgery/Procedures   Other        PS1120886

In [ ]:
# Combine rows in med_df based on petstablished_id
med_df = med_df.groupby('petstablished_id').agg(lambda x: ', '.join(x.dropna())).reset_index()

In [ ]:
len(med_df)

1089

In [ ]:
med_df.head(2)

petstablished_id  record_type                  subtype
0         PS101047  Medications   Alprazolam, Fluoxetine
1        PS1037920    Diagnosis

In [ ]:
# Change all entries in med_df to lowercase
med_df = med_df.applymap(lambda x: x.lower() if isinstance(x, str) else x)

C:\Users\runek\AppData\Local\Temp\ipykernel_7352\3127385508.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  med_df = med_df.applymap(lambda x: x.lower() if isinstance(x, str) else x)


In [ ]:
med_df.subtype.value_counts()

subtype
microchip implantation                                       373
                                                             321
dental prophylaxis                                           186
other                                                         63
medications                                                   13
                                                            ... 
dental prophylaxis, other, other, other                        1
medications, other, other, other                               1
dental prophylaxis, biopsy reports                             1
other, other, fecal evaluation, dental prophylaxis, other      1
radio graphs, dental prophylaxis                               1
Name: count, Length: 64, dtype: int64

We want to remove 'microchip' implantation records because that is standard and not impactful to the dog's health

In [ ]:
med_df = med_df[med_df.subtype != 'microchip implantation']

In [ ]:
len(med_df)

716

In [ ]:
# If you want to keep all rows in df even if there's no match in med_df
cleaned_df = pd.merge(df, med_df, on = 'petstablished_id', how = 'left')

In [ ]:
cleaned_df.head()

length_of_stay_days current_status pet_type pet_age petstablished_id  \
0                   43        adopted      dog  senior        ps1120886   
1                   24        adopted      dog  senior         ps333886   
2                   99        adopted      dog  senior          ps35514   
3                   31        adopted      dog   adult        ps1564518   
4                   50        adopted      dog   puppy        ps1357978   

    size        shedding coat_length temperament breed_type adoption_fee  \
0  small     no shedding         nan         nan        nan          400   
1  small             nan         nan         nan        nan          250   
2  small  sheds a little       short         nan  purebreed          200   
3  small     no shedding         nan         nan        nan          550   
4  large             nan         nan         nan        nan          465   

   gender is_mix shots_up_to_date spayed_neutered hypoallergenic housebroken  \
0  female    yes              yes             yes            yes         yes   
1  female     no              yes             yes            yes         yes   
2  female     no              yes             yes             no          no   
3    male     no              yes             yes       not sure          no   
4    male    yes              yes             yes       not sure    not sure   

  special_need gets_along_with_cats gets_along_with_dogs gets_along_with_kids  \
0           no                  yes                  yes                  yes   
1           no             not sure                  yes             not sure   
2           no             not sure                  yes                  yes   
3     not sure             not sure                  yes             not sure   
4     not sure             not sure             not sure             not sure   

  courtesy_post cross_post litter parent_of_litter          primary_breed  \
0            no         no     no               no   schnauzer (standard)   
1            no         no     no               no  schnauzer (miniature)   
2           yes         no     no               no              dachshund   
3            no         no     no               no                maltese   
4            no         no     no               no        black mouth cur   

         secondary_breed third_breed fourth_breed bully_breed popular_breed  \
0   jack russell terrier        None         None          no            no   
1                    NaN        None         None          no            no   
2                    NaN        None         None          no           yes   
3                    NaN        None         None          no            no   
4  shepherds (all types)        None         None          no            no   

  bonded_pair  age_total_months  sentiment_score         record_type  \
0          no             162.0              0.0  surgery/procedures   
1          no             170.0              0.0  surgery/procedures   
2          no             199.0              0.0                 NaN   
3          no              65.0              0.0                 NaN   
4          no              25.0              0.0                 NaN   

              subtype  
0               other  
1  dental prophylaxis  
2                 NaN  
3                 NaN  
4                 NaN

In [ ]:
cleaned_df.record_type.unique()

array(['surgery/procedures', nan, 'medications, surgery/procedures',
       'diagnosis, diagnosis', 'diagnosis', 'medications, diagnosis',
       'medications, medications, medications, medications',
       'surgery/procedures, medications', 'medications, medications',
       'medications', 'diagnosis, medications',
       'surgery/procedures, diagnosis',
       'surgery/procedures, surgery/procedures',
       'diagnosis, diagnosis, medications',
       'medications, surgery/procedures, surgery/procedures, diagnosis, surgery/procedures',
       'medications, medications, medications',
       'medications, medications, medications, medications, medications, medications, medications, medications, medications, medications, medications, medications',
       'surgery/procedures, medications, medications, medications',
       'surgery/procedures, surgery/procedures, medications',
       'diagnosis, medications, diagnosis',
       'diagnosis, surgery/procedures, diagnosis, medications',
     

In [ ]:
# Drop unneeded columns from cleaned_df
cleaned_df = cleaned_df.drop(['current_status', 'petstablished_id'], axis = 1)

In [ ]:
cleaned_df.head()

length_of_stay_days pet_type pet_age   size        shedding coat_length  \
0                   43      dog  senior  small     no shedding         nan   
1                   24      dog  senior  small             nan         nan   
2                   99      dog  senior  small  sheds a little       short   
3                   31      dog   adult  small     no shedding         nan   
4                   50      dog   puppy  large             nan         nan   

  temperament breed_type adoption_fee  gender is_mix shots_up_to_date  \
0         nan        nan          400  female    yes              yes   
1         nan        nan          250  female     no              yes   
2         nan  purebreed          200  female     no              yes   
3         nan        nan          550    male     no              yes   
4         nan        nan          465    male    yes              yes   

  spayed_neutered hypoallergenic housebroken special_need  \
0             yes            yes         yes           no   
1             yes            yes         yes           no   
2             yes             no          no           no   
3             yes       not sure          no     not sure   
4             yes       not sure    not sure     not sure   

  gets_along_with_cats gets_along_with_dogs gets_along_with_kids  \
0                  yes                  yes                  yes   
1             not sure                  yes             not sure   
2             not sure                  yes                  yes   
3             not sure                  yes             not sure   
4             not sure             not sure             not sure   

  courtesy_post cross_post litter parent_of_litter          primary_breed  \
0            no         no     no               no   schnauzer (standard)   
1            no         no     no               no  schnauzer (miniature)   
2           yes         no     no               no              dachshund   
3            no         no     no               no                maltese   
4            no         no     no               no        black mouth cur   

         secondary_breed third_breed fourth_breed bully_breed popular_breed  \
0   jack russell terrier        None         None          no            no   
1                    NaN        None         None          no            no   
2                    NaN        None         None          no           yes   
3                    NaN        None         None          no            no   
4  shepherds (all types)        None         None          no            no   

  bonded_pair  age_total_months  sentiment_score         record_type  \
0          no             162.0              0.0  surgery/procedures   
1          no             170.0              0.0  surgery/procedures   
2          no             199.0              0.0                 NaN   
3          no              65.0              0.0                 NaN   
4          no              25.0              0.0                 NaN   

              subtype  
0               other  
1  dental prophylaxis  
2                 NaN  
3                 NaN  
4                 NaN

In [ ]:
# Check number of records in cleaned_df
len(cleaned_df)

1978

In [ ]:
# Add a new column called 'had_medical' to describe whether the animal has needed medical attention
cleaned_df['had_medical'] = ''

In [ ]:
# If the record_type column contains data, put 'yes' in the had_medical column. Otherwise put 'no'
cleaned_df['had_medical'] = np.where(cleaned_df.record_type.notnull(), 'yes', 'no')

In [ ]:
cleaned_df.head()

length_of_stay_days pet_type pet_age   size        shedding coat_length  \
0                   43      dog  senior  small     no shedding         nan   
1                   24      dog  senior  small             nan         nan   
2                   99      dog  senior  small  sheds a little       short   
3                   31      dog   adult  small     no shedding         nan   
4                   50      dog   puppy  large             nan         nan   

  temperament breed_type adoption_fee  gender is_mix shots_up_to_date  \
0         nan        nan          400  female    yes              yes   
1         nan        nan          250  female     no              yes   
2         nan  purebreed          200  female     no              yes   
3         nan        nan          550    male     no              yes   
4         nan        nan          465    male    yes              yes   

  spayed_neutered hypoallergenic housebroken special_need  \
0             yes            yes         yes           no   
1             yes            yes         yes           no   
2             yes             no          no           no   
3             yes       not sure          no     not sure   
4             yes       not sure    not sure     not sure   

  gets_along_with_cats gets_along_with_dogs gets_along_with_kids  \
0                  yes                  yes                  yes   
1             not sure                  yes             not sure   
2             not sure                  yes                  yes   
3             not sure                  yes             not sure   
4             not sure             not sure             not sure   

  courtesy_post cross_post litter parent_of_litter          primary_breed  \
0            no         no     no               no   schnauzer (standard)   
1            no         no     no               no  schnauzer (miniature)   
2           yes         no     no               no              dachshund   
3            no         no     no               no                maltese   
4            no         no     no               no        black mouth cur   

         secondary_breed third_breed fourth_breed bully_breed popular_breed  \
0   jack russell terrier        None         None          no            no   
1                    NaN        None         None          no            no   
2                    NaN        None         None          no           yes   
3                    NaN        None         None          no            no   
4  shepherds (all types)        None         None          no            no   

  bonded_pair  age_total_months  sentiment_score         record_type  \
0          no             162.0              0.0  surgery/procedures   
1          no             170.0              0.0  surgery/procedures   
2          no             199.0              0.0                 NaN   
3          no              65.0              0.0                 NaN   
4          no              25.0              0.0                 NaN   

              subtype had_medical  
0               other         yes  
1  dental prophylaxis         yes  
2                 NaN          no  
3                 NaN          no  
4                 NaN          no

In [ ]:
cleaned_df.subtype.value_counts()

subtype
                                                                154
dental prophylaxis                                              145
other                                                            37
medications                                                      10
dental prophylaxis, dental prophylaxis                            9
dental prophylaxis, microchip implantation                        7
mass removal                                                      6
microchip implantation, other                                     5
other, dental prophylaxis                                         4
microchip implantation, dental prophylaxis                        4
medications, dental prophylaxis                                   3
other surgery                                                     3
dental prophylaxis, mass removal                                  3
medications, medications                                          3
other, other                            

In [ ]:
cleaned_df.length_of_stay_days.unique()

array([  43,   24,   99,   31,   50,   22,   90,   12,   28,  176,   61,
         27,   20,    8,   37,   30,   14,   10,   72,   80,   70,   57,
         56,   79,   16,   51,  146,   69,  181,  129,  318,   21,   36,
        606,  427,    3,   17,   46,    0,   83,   15,   48,  235,   25,
         40,   63,   76,   44,   74,   41,  444,   54,   35,  121,  119,
         18,  133,   55,   88,  173,   45,   73,    9,  154,    7,   42,
         89,   52,    6,  183,  122,   91,  187,  209,  354,  132,  170,
        138,   78,  115,    5,   53,   23,  107,   47,  130,   81,  126,
         84,    4,   29,  314,   34,   93,  163,  524,   95,   68,  105,
         19,   26,   62,   13,   92,  324,   39,   33,  104,   75,  207,
         66,   59,   11,   38,   67,  140,   60,  124,   49,  134,   64,
        453,  495,  117,   86,  303,  113,  120,  244,  229,  548,   32,
         96,  108,   58,  125, 1047,    2,  561,  111,  520,  155,  148,
        160,  226,  179,  103,  292,  137, 1020,  1

In [ ]:
cleaned_df.length_of_stay_days.max()

1807

In [ ]:
cleaned_df = cleaned_df[cleaned_df.bonded_pair != '$400']

In [ ]:
cleaned_df.to_csv('final_base_cleaned.csv', index = False)